In [87]:
import pandas as pd
import numpy as np


In [88]:
pd.__version__

'2.3.2'

Ans1. The version of pandas is 2.3.2

In [89]:
df = pd.read_csv('car_fuel_efficiency.csv')

In [90]:
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

In [91]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9704 entries, 0 to 9703
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   engine_displacement  9704 non-null   int64  
 1   num_cylinders        9222 non-null   float64
 2   horsepower           8996 non-null   float64
 3   vehicle_weight       9704 non-null   float64
 4   acceleration         8774 non-null   float64
 5   model_year           9704 non-null   int64  
 6   origin               9704 non-null   object 
 7   fuel_type            9704 non-null   object 
 8   drivetrain           9704 non-null   object 
 9   num_doors            9202 non-null   float64
 10  fuel_efficiency_mpg  9704 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 834.1+ KB


Ans2. There are total 9704 record counts


In [92]:
df['fuel_type'].unique()

array(['Gasoline', 'Diesel'], dtype=object)

Ans3. There are 2 fuel types in the dataset


In [93]:
df.isna().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

Ans4. 4 columns in the dataset has missing values

In [94]:
df['origin'].unique()

array(['Europe', 'USA', 'Asia'], dtype=object)

In [95]:
dfasia = df[df['origin']=='Asia']
dfasia.describe()

engine_displacement  num_cylinders   horsepower  vehicle_weight  \
count          3247.000000    3099.000000  3013.000000     3247.000000   
mean            200.052356       4.010326   150.303352     3006.292228   
std              49.287252       1.971199    30.257169      502.434626   
min              30.000000       0.000000    40.000000     1223.298226   
25%             170.000000       3.000000   130.000000     2665.518985   
50%             200.000000       4.000000   150.000000     2988.585737   
75%             230.000000       5.000000   171.000000     3345.355571   
max             370.000000      11.000000   245.000000     4661.144932   

       acceleration   model_year    num_doors  fuel_efficiency_mpg  
count   2933.000000  3247.000000  3072.000000          3247.000000  
mean      15.036754  2011.489375    -0.001628            14.973830  
std        2.488340     6.704607     1.051428             2.578345  
min        6.400000  2000.000000    -4.000000             6.886245  
25%       13.300000  2006.000000    -1.000000            13.252010  
50%       15.100000  2012.000000     0.000000            15.034670  
75%       16.700000  2017.000000     1.000000            16.726318  
max       22.700000  2023.000000     4.000000            23.759123

Ans5. 23.75 is the max fuel efficiency for cars in Asia

In [96]:
df['horsepower'].median()

149.0

In [97]:
df['horsepower'].value_counts().max()

142

In [98]:
df['horsepower'] = df['horsepower'].fillna(142)

In [99]:
df['horsepower'].median()

146.0

Ans6. The median of horsepowe was 149 with null values.
      The value with max counts is 142.
      After replacing null values with 142 the median changed to 146.
      Yes, it decreased.

In [100]:
X = dfasia[['vehicle_weight','model_year']].head(7)

In [101]:
XTX = X.T @ X

In [102]:
XTX_inv = np.linalg.inv(XTX)

In [104]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
w = XTX_inv @ X.T @ y
sum_w = w.sum()

In [105]:
print(f"Sum of elements of the result is {sum_w}")

Sum of elements of the result is 0.5187709081074023


Ans7. 0.518